File: lib/src/main/java/net/mm2d/upnp/ServiceParser.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp;

import net.mm2d.util.TextUtils;
import net.mm2d.util.XmlUtils;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import java.io.IOException;
import java.net.URL;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.xml.parsers.ParserConfigurationException;

/**
 * Serviceのパース処理。
 *
 * <p>Description XMLのダウンロード、パース、Builderへの値の設定をstatic methodで定義。
 *
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
final class ServiceParser {
    /**
     * SCPDURLからDescriptionを取得し、パースする。
     *
     * <p>可能であればKeepAliveを行う。
     *
     * @param client        通信に使用するHttpClient
     * @param deviceBuilder DeviceのBuilder
     * @param builder       ServiceのBuilder
     * @throws IOException                  通信エラー
     * @throws SAXException                 XMLパースエラー
     * @throws ParserConfigurationException 実装が使用できないかインスタンス化できない
     */
    static void loadDescription(
            @Nonnull final HttpClient client,
            @Nonnull final DeviceImpl.Builder deviceBuilder,
            @Nonnull final ServiceImpl.Builder builder)
            throws IOException, SAXException, ParserConfigurationException {
        final String scpdUrl = builder.getScpdUrl();
        if (scpdUrl == null) {
            throw new IOException();
        }
        final String baseUrl = deviceBuilder.getBaseUrl();
        final int scopeId = deviceBuilder.getSsdpMessage().getScopeId();
        final URL url = Http.makeAbsoluteUrl(baseUrl, scpdUrl, scopeId);
        final String description = client.downloadString(url);
        if (TextUtils.isEmpty(description)) {
            // 空であっても必須パラメータはそろっているため正常として扱う。
            return;
        }
        builder.setDescription(description);
        final Document doc = XmlUtils.newDocument(true, description);
        parseActionList(builder, doc.getElementsByTagName("action"));
        parseStateVariableList(builder, doc.getElementsByTagName("stateVariable"));
    }

    private static void parseActionList(
            @Nonnull final ServiceImpl.Builder builder,
            @Nonnull final NodeList nodeList) {
        for (int i = 0; i < nodeList.getLength(); i++) {
            builder.addActionBuilder(parseAction((Element) nodeList.item(i)));
        }
    }

    private static void parseStateVariableList(
            @Nonnull final ServiceImpl.Builder builder,
            @Nonnull final NodeList nodeList) {
        for (int i = 0; i < nodeList.getLength(); i++) {
            builder.addStateVariable(parseStateVariable((Element) nodeList.item(i)));
        }
    }

    @Nullable
    private static String getTagName(@Nonnull final Node node) {
        if (node.getNodeType() != Node.ELEMENT_NODE) {
            return null;
        }
        return node.getLocalName();
    }

    @Nonnull
    private static ActionImpl.Builder parseAction(@Nonnull final Element element) {
        final ActionImpl.Builder builder = new ActionImpl.Builder();
        Node node = element.getFirstChild();
        for (; node != null; node = node.getNextSibling()) {
            final String tag = getTagName(node);
            if (TextUtils.equals(tag, "name")) {
                builder.setName(node.getTextContent());
            } else if (TextUtils.equals(tag, "argumentList")) {
                for (Node c = node.getFirstChild(); c != null; c = c.getNextSibling()) {
                    if (TextUtils.equals(getTagName(c), "argument")) {
                        builder.addArgumentBuilder(parseArgument((Element) c));
                    }
                }
            }
        }
        return builder;
    }

    @Nonnull
    private static ArgumentImpl.Builder parseArgument(@Nonnull final Element element) {
        final ArgumentImpl.Builder builder = new ArgumentImpl.Builder();
        Node node = element.getFirstChild();
        for (; node != null; node = node.getNextSibling()) {
            final String tag = getTagName(node);
            if (TextUtils.isEmpty(tag)) {
                continue;
            }
            final String value = node.getTextContent();
            setField(builder, tag, value);
        }
        return builder;
    }

    @SuppressWarnings("IfCanBeSwitch")
    private static void setField(
            @Nonnull final ArgumentImpl.Builder builder,
            @Nonnull final String tag,
            @Nonnull final String value) {
        if ("name".equals(tag)) {
            builder.setName(value);
        } else if ("direction".equals(tag)) {
            builder.setDirection(value);
        } else if ("relatedStateVariable".equals(tag)) {
            builder.setRelatedStateVariableName(value);
        }
    }

    @Nonnull
    private static StateVariable parseStateVariable(@Nonnull final Element element) {
        final StateVariableImpl.Builder builder = new StateVariableImpl.Builder()
                .setSendEvents(element.getAttribute("sendEvents"))
                .setMulticast(element.getAttribute("multicast"));
        Node node = element.getFirstChild();
        for (; node != null; node = node.getNextSibling()) {
            final String tag = getTagName(node);
            if (TextUtils.isEmpty(tag)) {
                continue;
            }
            if ("name".equals(tag)) {
                builder.setName(node.getTextContent());
            } else if ("dataType".equals(tag)) {
                builder.setDataType(node.getTextContent());
            } else if ("defaultValue".equals(tag)) {
                builder.setDefaultValue(node.getTextContent());
            } else if ("allowedValueList".equals(tag)) {
                parseAllowedValueList(builder, (Element) node);
            } else if ("allowedValueRange".equals(tag)) {
                parseAllowedValueRange(builder, (Element) node);
            }
        }
        return builder.build();
    }

    private static void parseAllowedValueList(
            @Nonnull final StateVariableImpl.Builder builder,
            @Nonnull final Element element) {
        Node node = element.getFirstChild();
        for (; node != null; node = node.getNextSibling()) {
            final String tag = getTagName(node);
            if ("allowedValue".equals(tag)) {
                builder.addAllowedValue(node.getTextContent());
            }
        }
    }

    private static void parseAllowedValueRange(
            @Nonnull final StateVariableImpl.Builder builder,
            @Nonnull final Element element) {
        Node node = element.getFirstChild();
        for (; node != null; node = node.getNextSibling()) {
            final String tag = getTagName(node);
            if (TextUtils.isEmpty(tag)) {
                continue;
            }
            final String value = node.getTextContent();
            setField(builder, tag, value);
        }
    }

    @SuppressWarnings("IfCanBeSwitch")
    private static void setField(
            @Nonnull final StateVariableImpl.Builder builder,
            @Nonnull final String tag,
            @Nonnull final String value) {
        if ("step".equals(tag)) {
            builder.setStep(value);
        } else if ("minimum".equals(tag)) {
            builder.setMinimum(value);
        } else if ("maximum".equals(tag)) {
            builder.setMaximum(value);
        }
    }

    // インスタンス化禁止
    private ServiceParser() {
    }
}

File: mmupnp/src/main/java/net/mm2d/upnp/internal/parser/ServiceParser.kt
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.parser

import net.mm2d.upnp.Http
import net.mm2d.upnp.SingleHttpClient
import net.mm2d.upnp.StateVariable
import net.mm2d.upnp.internal.impl.ActionImpl
import net.mm2d.upnp.internal.impl.ArgumentImpl
import net.mm2d.upnp.internal.impl.DeviceImpl
import net.mm2d.upnp.internal.impl.ServiceImpl
import net.mm2d.upnp.internal.impl.StateVariableImpl
import net.mm2d.xml.node.XmlElement
import net.mm2d.xml.parser.XmlParser
import java.io.IOException

/**
 * Parser for Service.
 *
 * Download Description XML, parse it, set value to builder.
 *
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
internal object ServiceParser {
    private val deviceTypesThatAllowIoException = setOf(
        // Some DIAL devices return XML, so try the download.
        // But some DIAL devices return 404, so allow it.
        "urn:dial-multiscreen-org:device:dial:1",
        // Basic devices are allowed to have no Service.
        // However, some devices have a Service but return 404, so allow it.
        "urn:schemas-upnp-org:device:Basic:1",
    )

    /**
     * Download Description from SCPDURL and parse it.
     *
     * KeepAlive, if possible.
     *
     * @param client HttpClient
     * @param deviceBuilder DeviceのBuilder
     * @param builder ServiceのBuilder
     * @throws IOException if an I/O error occurs.
     */
    @Throws(IOException::class)
    fun loadDescription(client: SingleHttpClient, deviceBuilder: DeviceImpl.Builder, builder: ServiceImpl.Builder) {
        val scpdUrl = builder.getScpdUrl() ?: throw IOException("scpdUrl is null")
        val baseUrl = deviceBuilder.getBaseUrl()
        val scopeId = deviceBuilder.getSsdpMessage().scopeId
        val url = Http.makeAbsoluteUrl(baseUrl, scpdUrl, scopeId)
        val description = try {
            client.downloadString(url)
        } catch (e: IOException) {
            if (!deviceTypesThatAllowIoException.contains(deviceBuilder.getDeviceType())) {
                throw e
            }
            // Allow certain "ill-behaved" devices
            ""
        }
        if (description.isEmpty()) {
            // 空であっても必須パラメータはそろっているため正常として扱う。
            return
        }
        builder.setDescription(description)
        val rootElement = XmlParser.parse(description) ?: return

        rootElement.childElements
            .filter { it.localName == "actionList" }
            .flatMap { it.childElements }
            .filter { it.localName == "action" }
            .forEach { builder.addActionBuilder(parseAction(it)) }
        rootElement.childElements
            .filter { it.localName == "serviceStateTable" }
            .flatMap { it.childElements }
            .filter { it.localName == "stateVariable" }
            .forEach { builder.addStateVariable(parseStateVariable(it)) }
    }

    private fun parseAction(element: XmlElement): ActionImpl.Builder {
        val builder = ActionImpl.Builder()
        element.childElements
            .forEach {
                when (it.localName) {
                    "name" -> builder.setName(it.value)
                    "argumentList" -> parseArgumentList(builder, it)
                }
            }
        return builder
    }

    private fun parseArgumentList(builder: ActionImpl.Builder, element: XmlElement) {
        element.childElements
            .filter { it.localName == "argument" }
            .forEach { builder.addArgumentBuilder(parseArgument(it)) }
    }

    private fun parseArgument(element: XmlElement): ArgumentImpl.Builder {
        val builder = ArgumentImpl.Builder()
        element.childElements
            .forEach { builder.setField(it.localName, it.value) }
        return builder
    }

    private fun ArgumentImpl.Builder.setField(tag: String, value: String) {
        when (tag) {
            "name" ->
                setName(value)
            "direction" ->
                setDirection(value)
            "relatedStateVariable" ->
                setRelatedStateVariableName(value)
        }
    }

    private fun parseStateVariable(element: XmlElement): StateVariable {
        val builder = StateVariableImpl.Builder()
        builder.setSendEvents(element.getAttributeValue("sendEvents"))
        builder.setMulticast(element.getAttributeValue("multicast"))
        element.childElements.forEach {
            when (it.localName) {
                "name" ->
                    builder.setName(it.value)
                "dataType" ->
                    builder.setDataType(it.value)
                "defaultValue" ->
                    builder.setDefaultValue(it.value)
                "allowedValueList" ->
                    parseAllowedValueList(builder, it)
                "allowedValueRange" ->
                    parseAllowedValueRange(builder, it)
            }
        }
        return builder.build()
    }

    private fun parseAllowedValueList(builder: StateVariableImpl.Builder, element: XmlElement) {
        element.childElements
            .filter { it.localName == "allowedValue" }
            .forEach { builder.addAllowedValue(it.value) }
    }

    private fun parseAllowedValueRange(builder: StateVariableImpl.Builder, element: XmlElement) {
        element.childElements
            .forEach { builder.setField(it.localName, it.value) }
    }

    private fun StateVariableImpl.Builder.setField(tag: String, value: String) {
        when (tag) {
            "step" ->
                setStep(value)
            "minimum" ->
                setMinimum(value)
            "maximum" ->
                setMaximum(value)
        }
    }
}
